File: python/tav2/memory.py
"""混合检索：滚动摘要 + 世界书关键词激活 + 锁定术语 + TM few-shot + 可选向量兜底。"""

from __future__ import annotations

import json
from dataclasses import asdict
from dataclasses import dataclass, field
from typing import Any

from tav2.llm import embed_texts
from tav2.models import Scene, UnderstandingRecord
from tav2.scanning import target_chars
from tav2.worldbook import activate


@dataclass
class MemoryPack:
    summary: str = ""
    main_summary: str = ""  # 分支场景携带的主线摘要（合并点上下文）
    constants: list[dict[str, Any]] = field(default_factory=list)
    lore_hits: list[dict[str, Any]] = field(default_factory=list)
    glossary: list[tuple[str, str, str]] = field(default_factory=list)
    few_shot: list[tuple[str, str]] = field(default_factory=list)
    recent_understandings: list[UnderstandingRecord] = field(default_factory=list)
    vector_hits: list[dict[str, Any]] = field(default_factory=list)


def build_memory_pack(
    db: Any,
    llm: Any,
    cfg: dict[str, Any],
    scene: Scene,
    unit_sources: list[str],
    main_branch: str | None = None,
) -> MemoryPack:
    """为场景组装记忆包。"""

    main_branch = main_branch or "main"
    pack = MemoryPack()
    pack.summary = db.get_summary(scene.branch)
    if scene.branch != main_branch:
        # 合并点/分支场景：额外携带主线摘要，保证分支前文上下文不丢
        pack.main_summary = db.get_summary(main_branch)
    worldbook = db.load_worldbook()
    joined = "\n".join(unit_sources)
    pack.constants, pack.lore_hits = activate(worldbook, joined)

    locked = db.locked_terms()
    pack.glossary = [
        (t["source"], t["target"], t.get("category") or "")
        for t in locked
        if _term_hits(t["source"], joined)
    ]
    pack.recent_understandings = db.recent_understandings(scene.branch, limit=3)
    pack.few_shot = _few_shot(db, unit_sources, int(cfg.get("context", {}).get("few_shot_pairs", 6)))

    if cfg.get("memory", {}).get("vector_enabled"):
        candidates = _index_candidates(worldbook, pack.recent_understandings)
        pack.vector_hits = _vector_hits(cfg, llm, candidates, unit_sources)
    return pack


def _term_hits(source: str, joined: str) -> bool:
    import re

    if not source or len(source) < 2:
        return False
    if re.search(r"[A-Za-z]", source):
        return (
            re.search(
                r"(?<![A-Za-z0-9_])" + re.escape(source) + r"(?![A-Za-z0-9_])",
                joined,
            )
            is not None
        )
    return source in joined


def _few_shot(db: Any, unit_sources: list[str], max_pairs: int) -> list[tuple[str, str]]:
    pairs: list[tuple[str, str]] = []
    for source in unit_sources:
        translation = db.tm_get(source)
        if translation and translation != source:
            pairs.append((source, translation))
        if len(pairs) >= max_pairs:
            break
    return pairs


def _vector_hits(
    cfg: dict[str, Any],
    llm: Any,
    candidates: list[dict[str, Any]],
    unit_sources: list[str],
) -> list[dict[str, Any]]:
    """向量检索兜底：对索引候选（世界书 + 最近理解记录）做 top-k 召回。

    embedding 未配置/调用失败时静默降级返回 []。
    """

    if not candidates or not unit_sources:
        return []

    def embed(texts: list[str]) -> list[list[float]]:
        try:
            return embed_texts(cfg, texts) or []
        except Exception:
            return []

    query_vecs = embed(["\n".join(unit_sources)])
    if not query_vecs:
        return []
    index = build_vector_index(candidates, embed)
    if not index:
        return []
    top_k = int(cfg.get("memory", {}).get("top_k", 3))
    return vector_search(query_vecs[0], index, top_k)


def _index_candidates(
    worldbook: list[dict[str, Any]],
    recent_understandings: list[Any],
) -> list[dict[str, Any]]:
    """索引候选 = 世界书条目 + 最近理解记录（序列化为可检索文本）。"""

    candidates: list[dict[str, Any]] = []
    for entry in worldbook:
        candidates.append(
            {
                "kind": str(entry.get("kind") or "lore"),
                "title": str(entry.get("title") or ""),
                "keywords": list(entry.get("keywords") or []),
                "content": str(entry.get("content") or ""),
                "source_refs": list(entry.get("source_refs") or []),
            }
        )
    for rec in recent_understandings:
        text = json.dumps(
            {"state": rec.scene_state, "threads": [asdict(t) for t in rec.threads]},
            ensure_ascii=False,
        )
        candidates.append(
            {
                "kind": "understanding",
                "title": f"理解:{rec.scene_id}",
                "keywords": [],
                "content": text,
                "source_refs": [],
            }
        )
    return candidates


def build_vector_index(
    entries: list[dict[str, Any]],
    embed_fn,
) -> list[tuple[dict[str, Any], list[float]]]:
    """把条目批量向量化，返回 [(entry, vec)]（失败/空向量条目跳过）。"""

    if not entries:
        return []
    texts = [
        str(e.get("content") or "") or str(e.get("title") or "")
        for e in entries
    ]
    vecs = embed_fn(texts)
    if not vecs:
        return []
    out: list[tuple[dict[str, Any], list[float]]] = []
    for entry, vec in zip(entries, vecs):
        if vec:
            out.append((entry, vec))
    return out


def vector_search(
    query_vec: list[float],
    index: list[tuple[dict[str, Any], list[float]]],
    top_k: int,
    min_score: float = 0.5,
) -> list[dict[str, Any]]:
    """余弦相似度 top-k 召回（高于 min_score 才保留）。"""

    scored: list[tuple[float, dict[str, Any]]] = []
    for entry, vec in index:
        score = _cosine(query_vec, vec)
        if score > min_score:
            scored.append((score, entry))
    scored.sort(key=lambda x: -x[0])
    return [entry for _score, entry in scored[: max(0, top_k)]]


def _cosine(a: list[float], b: list[float]) -> float:
    if not a or not b or len(a) != len(b):
        return 0.0
    dot = sum(x * y for x, y in zip(a, b))
    na = sum(x * x for x in a) ** 0.5
    nb = sum(x * x for x in b) ** 0.5
    if na == 0 or nb == 0:
        return 0.0
    return dot / (na * nb)

File: src/engine/memory.ts
/**
 * 混合检索（与 Python tav2/memory.py 对齐）：
 * 摘要 + 世界书关键词激活 + 锁定术语 + TM few-shot + 最近理解 + 向量兜底。
 * embedding 未配置或调用失败时静默降级为 []。
 */

import type { EngineConfig } from './config'
import type { ProjectDB } from './db'
import type { Scene, UnderstandingRecord } from './models'

export type MemoryEntry = Record<string, unknown>

/** 向量索引：候选条目 + 其嵌入向量。 */
export type VectorIndex = Array<[MemoryEntry, number[]]>

/** 文本向量化函数（可注入，便于离线测试）。 */
export type EmbedFn = (texts: string[]) => Promise<number[][]>

export interface MemoryPack {
  summary: string
  /** 分支场景携带的主线摘要（合并点上下文） */
  mainSummary: string
  constants: MemoryEntry[]
  loreHits: MemoryEntry[]
  glossary: Array<[string, string, string]>
  fewShot: Array<[string, string]>
  recentUnderstandings: UnderstandingRecord[]
  /** 向量召回（embedding 未配置时为空） */
  vectorHits: MemoryEntry[]
}

function escapeRegex(text: string): string {
  return text.replace(/[.*+?^${}()|[\]\\]/g, '\\$&')
}

function termHits(source: string, joined: string): boolean {
  if (!source || source.length < 2) return false
  if (/[A-Za-z]/.test(source)) {
    return new RegExp(`(?<![A-Za-z0-9_])${escapeRegex(source)}(?![A-Za-z0-9_])`).test(joined)
  }
  return joined.includes(source)
}

/** 按原文关键词词边界激活条目。返回 (常驻, 命中)。 */
export function activateEntries(entries: MemoryEntry[], text: string): [MemoryEntry[], MemoryEntry[]] {
  const constants: MemoryEntry[] = []
  const hits: MemoryEntry[] = []
  for (const entry of entries) {
    if (entry.kind === 'constant') {
      constants.push(entry)
      continue
    }
    for (const raw of (entry.keywords as unknown[] | undefined) ?? []) {
      const kw = String(raw ?? '').trim()
      if (!kw) continue
      if (/[A-Za-z]/.test(kw)) {
        if (new RegExp(`(?<![A-Za-z0-9_])${escapeRegex(kw)}(?![A-Za-z0-9_])`, 'i').test(text)) {
          hits.push(entry)
          break
        }
      } else if (text.includes(kw)) {
        hits.push(entry)
        break
      }
    }
  }
  return [constants, hits]
}

function fewShot(db: ProjectDB, unitSources: string[], maxPairs: number): Array<[string, string]> {
  const pairs: Array<[string, string]> = []
  for (const source of unitSources) {
    const translation = db.tmGet(source)
    if (translation && translation !== source) pairs.push([source, translation])
    if (pairs.length >= maxPairs) break
  }
  return pairs
}

/** 向量索引候选 = 世界书条目 + 最近理解记录（序列化为可检索文本）。 */
export function _indexCandidates(
  worldbook: MemoryEntry[],
  recentUnderstandings: UnderstandingRecord[],
): MemoryEntry[] {
  const candidates: MemoryEntry[] = []
  for (const entry of worldbook) {
    candidates.push({
      kind: String(entry.kind ?? 'lore'),
      title: String(entry.title ?? ''),
      keywords: Array.isArray(entry.keywords) ? [...(entry.keywords as unknown[])] : [],
      content: String(entry.content ?? ''),
      source_refs: Array.isArray(entry.source_refs) ? [...(entry.source_refs as unknown[])] : [],
    })
  }
  for (const rec of recentUnderstandings) {
    const text = JSON.stringify(
      {
        state: rec.scene_state,
        threads: rec.threads.map((t) => ({
          id: t.id,
          kind: t.kind,
          text: t.text,
          scenes_since: t.scenes_since,
        })),
      },
      null,
      0,
    )
    candidates.push({
      kind: 'understanding',
      title: `理解:${rec.scene_id}`,
      keywords: [],
      content: text,
      source_refs: [],
    })
  }
  return candidates
}

/** 把候选条目批量向量化，返回 [(entry, vec)]（失败/空向量条目跳过）。 */
export async function buildVectorIndex(
  entries: MemoryEntry[],
  embed: EmbedFn,
): Promise<VectorIndex> {
  if (entries.length === 0) return []
  const texts = entries.map((e) => String(e.content ?? '') || String(e.title ?? ''))
  let vecs: number[][]
  try {
    vecs = (await embed(texts)) ?? []
  } catch {
    return []
  }
  if (!vecs || vecs.length === 0) return []
  const out: VectorIndex = []
  for (let i = 0; i < entries.length; i++) {
    const vec = vecs[i]
    if (vec && vec.length > 0) out.push([entries[i]!, vec])
  }
  return out
}

function cosine(a: number[], b: number[]): number {
  if (!a || !b || a.length !== b.length) return 0
  let dot = 0
  let na = 0
  let nb = 0
  for (let i = 0; i < a.length; i++) {
    dot += a[i]! * b[i]!
    na += a[i]! * a[i]!
    nb += b[i]! * b[i]!
  }
  if (na === 0 || nb === 0) return 0
  return dot / (Math.sqrt(na) * Math.sqrt(nb))
}

/** 余弦相似度 top-k 召回（严格高于 min_score 才保留，与 Python 一致）。 */
export function vectorSearch(
  queryVec: number[],
  index: VectorIndex,
  topK: number,
  minScore = 0.5,
): MemoryEntry[] {
  const scored: Array<{ score: number; entry: MemoryEntry }> = []
  for (const [entry, vec] of index) {
    const score = cosine(queryVec, vec)
    if (score > minScore) scored.push({ score, entry })
  }
  scored.sort((a, b) => b.score - a.score)
  return scored.slice(0, Math.max(0, topK)).map((s) => s.entry)
}

/** 按引擎配置调用 OpenAI 兼容 /embeddings；未配置/无 key/失败静默返回 []。 */
export async function embedTexts(cfg: EngineConfig, texts: string[]): Promise<number[][]> {
  if (!cfg.memory.vectorEnabled) return []
  const model = cfg.memory.embeddingModel.trim()
  if (!model || texts.length === 0) return []
  const key = cfg.llm.apiKey
    || (cfg.llm.apiKeyEnv ? process.env[cfg.llm.apiKeyEnv] ?? '' : '')
    || process.env.TRANSLATE_AGENT_API_KEY
    || ''
  if (!key) return []
  try {
    const response = await fetch(`${cfg.llm.baseUrl.replace(/\/$/, '')}/embeddings`, {
      method: 'POST',
      headers: {
        Authorization: `Bearer ${key}`,
        'Content-Type': 'application/json',
      },
      body: JSON.stringify({ model, input: texts }),
      signal: AbortSignal.timeout(Math.max(1, cfg.llm.timeout) * 1000),
    })
    if (!response.ok) return []
    const data = await response.json() as {
      data?: Array<{ embedding?: number[] }>
    }
    return (data.data ?? [])
      .map((item) => item.embedding ?? [])
      .filter((vec) => Array.isArray(vec) && vec.length > 0)
  } catch {
    return []
  }
}

async function vectorHits(
  cfg: EngineConfig,
  candidates: MemoryEntry[],
  unitSources: string[],
): Promise<MemoryEntry[]> {
  if (candidates.length === 0 || unitSources.length === 0) return []
  const queryVecs = await embedTexts(cfg, [unitSources.join('\n')])
  if (queryVecs.length === 0) return []
  const index = await buildVectorIndex(candidates, (texts) => embedTexts(cfg, texts))
  if (index.length === 0) return []
  return vectorSearch(queryVecs[0]!, index, cfg.memory.topK)
}

/** 为场景组装记忆包（M4：向量兜底已启用，失败静默降级）。 */
export async function buildMemoryPack(
  db: ProjectDB,
  cfg: EngineConfig,
  scene: Scene,
  unitSources: string[],
  mainBranch = 'main',
): Promise<MemoryPack> {
  const pack: MemoryPack = {
    summary: db.getSummary(scene.branch),
    mainSummary: '',
    constants: [],
    loreHits: [],
    glossary: [],
    fewShot: [],
    recentUnderstandings: [],
    vectorHits: [],
  }
  if (scene.branch !== mainBranch) {
    pack.mainSummary = db.getSummary(mainBranch)
  }
  const worldbook = db.loadWorldbook()
  const joined = unitSources.join('\n')
  ;[pack.constants, pack.loreHits] = activateEntries(worldbook, joined)

  const locked = db.lockedTerms()
  pack.glossary = locked
    .filter((t) => termHits(String(t.source ?? ''), joined))
    .map((t) => [String(t.source ?? ''), String(t.target ?? ''), String(t.category ?? '')])

  pack.recentUnderstandings = db.recentUnderstandings(scene.branch, 3)
  pack.fewShot = fewShot(db, unitSources, cfg.context.fewShotPairs)
  const candidates = _indexCandidates(worldbook, pack.recentUnderstandings)
  pack.vectorHits = await vectorHits(cfg, candidates, unitSources)
  return pack
}
